deepchem/molnet/load_function/load_dataset_template.py

```diff
@@ -0,0 +1,202 @@
+    Try to reload dataset from disk if already downloaded. Save to disk
+    after featurizing.
+  data_dir: str, optional
+    Path to datasets.
+  save_dir: str, optional
+    Path to featurized datasets.
+  **kwargs: optional arguments to methods of featurizers, transformers, and
+  splitters.
+
+  Returns
+  -------
+  tasks, datasets, transformers : tuple
+    tasks : list
+      Column names corresponding to machine learning target variables.
+    datasets : tuple
+      train, validation, test splits of data as
+      ``deepchem.data.datasets.Dataset`` instances.
+    transformers : list
+      ``deepchem.trans.transformers.Transformer`` instances applied
+      to dataset.
+
+  References
+  ----------
+  MLA style references for this dataset. E.g.
+    Wu, Zhenqin et al. "MoleculeNet: a benchmark for molecular
+      machine learning." Chemical Science, vol. 9, 2018, 
+      pp. 513-530, 10.1039/c7sc02664a.
+
+    Last, First et al. "Article title." Journal name, vol. #,
+      no. #, year, pp. page range, DOI. 
+
+  Examples
+  --------
+  >> import deepchem as dc
+  >> tasks, datasets, transformers = dc.molnet.load_tox21(reload=False)
+  >> train_dataset, val_dataset, test_dataset = datasets
+  >> n_tasks = len(tasks)
+  >> n_features = train_dataset.get_data_shape()[0]
+  >> model = dc.models.MultitaskClassifier(n_tasks, n_features)
+
+  """
+
+  # Warning message about this template
+  raise ValueError("""
+    This is a template function and it doesn't do anything!
+    Use this function as a reference when implementing new
+    loaders for MoleculeNet datasets.
+    """)
+
+  # Featurize mydataset
+  logger.info("About to featurize mydataset.")
+  my_tasks = ["task1", "task2", "task3"]  # machine learning targets
+
+  # Get DeepChem data directory if needed
+  if data_dir is None:
+    data_dir = DEFAULT_DIR
+  if save_dir is None:
+    save_dir = DEFAULT_DIR
+
+  # Check for str args to featurizer, splitter, and transformers
+  if isinstance(featurizer, str):
+    featurizer = DEFAULT_FEATURIZERS[featurizer]
+  if isinstance(splitter, str):
+    splitter = DEFAULT_SPLITTERS[splitter]
+  transformers = [
+      DEFAULT_TRANSFORMERS[t] if isinstance(t, str) else t for t in transformers
+  ]
+
+  # Reload from disk
+  if reload:
+    featurizer_name = str(featurizer.__class__.__name__)
+    splitter_name = str(splitter.__class__.__name__)
+    save_folder = os.path.join(save_dir, "mydataset-featurized",
+                               featurizer_name, splitter_name)
+
+    loaded, all_dataset, transformers = deepchem.utils.save.load_dataset_from_disk(
+        save_folder)
+    if loaded:
+      return my_tasks, all_dataset, transformers
+
+  # 3D coordinate featurizers, e.g. 'CoulombMatrix' or 'MP'
+  # For crystal structures, replace with json_featurizers
+  sdf_featurizers = []  # type: List[Featurizer]
+
+  # If featurizer requires a non-CSV file format, load .tar.gz file
+  if featurizer in sdf_featurizers:
```

Comment:
This is an open question since I don't have a good solution to this, but is there a good way to generalize this to other types of loaders? If we have an image dataset, we might want to use `ImageLoader` for example. I don't know how to create a general template for this though.

Reply:
I can rework this to emphasize that a loader should have a list of accepted featurizers, and then the function needs to choose the appropriate `DataLoader` corresponding to any chosen featurization.

Reply:
I've abstracted this bit of the template to be for a generic `DataLoader` and specified that the user should implement any loaders as needed depending on the dataset file type and supported featurizations.